builtin/method_str.py

```diff
@@ -488,38 +488,89 @@ def __init__(self):
         """
         string = rd.PosStr()
-        sep = rd.PosStr()
+
+        string_sep = None  # type: str
+        eggex_sep = None  # type: value.Eggex
+
+        sep = rd.PosValue()
+        with tagswitch(sep) as case:
+            if case(value_e.Eggex):
+                eggex_sep_ = cast(value.Eggex, sep)
+                eggex_sep = eggex_sep_
+
+            elif case(value_e.Str):
+                string_sep_ = cast(value.Str, sep)
+                string_sep = string_sep_.s
+
+            else:
+                raise error.TypeErr(sep, 'expected separator to be Eggex or Str',
+                                    rd.LeftParenToken())
+
         count = mops.BigTruncate(rd.NamedInt("count", -1))
         rd.Done()
 
-        if len(sep) == 0:
-            raise error.Structured(3, "sep must be non-empty", rd.LeftParenToken())
-
         if len(string) == 0:
             return value.List([])
 
-        cursor = 0
-        chunks = []  # type: List[value_t]
-        while cursor < len(string) and count != 0:
-            next = string.find(sep, cursor)
-            if next == -1:
-                break
+        if string_sep is not None:
+            if len(string_sep) == 0:
+                raise error.Structured(3, "separator must be non-empty",
+                                       rd.LeftParenToken())
 
-            chunks.append(value.Str(string[cursor:next]))
-            cursor = next + len(sep)
-            count -= 1
+            cursor = 0
+            chunks = []  # type: List[value_t]
+            while cursor < len(string) and count != 0:
+                next = string.find(string_sep, cursor)
+                if next == -1:
+                    break
+
+                chunks.append(value.Str(string[cursor:next]))
+                cursor = next + len(string_sep)
+                count -= 1
 
-        if cursor == len(string):
-            # An instance of sep was against the end of the string
-            chunks.append(value.Str(""))
-        else:
             chunks.append(value.Str(string[cursor:]))
 
-        return value.List(chunks)
+            return value.List(chunks)
+
+        if eggex_sep is not None:
+            if '\0' in string:
+                raise error.Structured(
+                    3, "cannot split a string with a NUL byte",
+                    rd.LeftParenToken())
+
+            regex = regex_translate.AsPosixEre(eggex_sep)
+            cflags = regex_translate.LibcFlags(eggex_sep.canonical_flags)
+
+            cursor = 0
+            chunks = []
+            while cursor < len(string) and count != 0:
+                m = libc.regex_search(regex, cflags, string, 0, cursor)
+                if m is None:
+                    break
+
+                start = m[0]
+                end = m[1]
+                if start == end:
+                    raise error.Structured(
+                        3,
+                        "eggex separators should never match the empty string",
```

Comment:
This error reminds me that it would be nice if it were possible to take a STRING regex, in  ERE syntax

The ~ operator in Oils can do that, e.g. `if ('mystr' ~ 'myregex+')`

But the signature doesn't really allow that

---

But I'm OK just leaving this out for now ... I think it is possible to just write your own ERE string split in pure YSH

It would be nice to handle this, but we can put it off until someone asks I think   ... 

(I want to get through Hay/Modules/pure functions before going to deep, and we definitely need this split(), and split() on whitespace)